### DISCUSSION TITLE
How can I use RTSP H265?

### DISCUSSION THREAD

**Discussion by whitetiger21022014** [nanosonde/sip2rtsp#12] [Q&A]
Hello,
Could you let me know how can I use rtsp H265 (I don't need to use ffmpeg tool to transcode from H265 to H264)?
Thanks you

**Comment by nanosonde** [comment]
Hi @whitetiger21022014 !
So what you want is to do the same thing, but with H265 instead of H264? I assume that your camera already provides an H265 stream, right?

The `launch_string` config.yaml parameter for H264 looks like this ith re-encoding:

`rtspsrc location=rtsp://10.10.10.10:8554/gartentor_h264_ext name=rtspsrc ! decodebin ! videoconvert ! videoscale ! videorate ! video/x-raw,format=I420,width=1280,height=720,framerate=15/1 ! x264enc bitrate=2000 speed-preset=superfast tune=zerolatency key-int-max=60 ! rtph264pay config-interval=1 name=pay0 pt=96 pulsesrc device="BaresipSpeakerInput" do-timestamp=true ! queue ! audioconvert ! audioresample ! audio/x-raw,format=S16LE,channels=1,rate=8000 ! opusenc ! rtpopuspay name=pay1`

This `launch_string` will select only the video stream from the RTSP camera. The `decodebin` element would automatically select the required decoder. Later the stream is re-encoded to H264 using `x264enc`. 
Audio is selected from the SIP client's speaker and encoded to OPUS.

First I would suggest that you experiment with a gstreamer pipeline to get a working H265 video stream without audio from the gstreamer RTSP server. 
I presume that you do not want to re-encode (with [x265enc](https://gstreamer.freedesktop.org/documentation/x265/index.html?gi-language=python)) the H265 stream, right?

`rtspsrc location=rtsp://<IP>:8554/h265_stream name=rtspsrc ! parsebin ! rtph265pay config-interval=1 name=pay0 pt=96`

Later one could add audio again:
`rtspsrc location=rtsp://<IP>:8554/h265_stream name=rtspsrc ! parsebin ! rtph265pay config-interval=1 name=pay0 pt=96 pulsesrc device="BaresipSpeakerInput" do-timestamp=true ! queue ! audioconvert ! audioresample ! audio/x-raw,format=S16LE,channels=1,rate=8000 ! opusenc ! rtpopuspay name=pay1`

**Reply by whitetiger21022014** [reply]
"assume that your camera already provides an H265 stream, right?" Yes

**Comment by nanosonde** [comment]
Forgot one thing to add: the RTSP server is independent from the ONVIF server. The ONVIF server just describes what the RTSP server will deliver. The ONVIF server included in this repo just has some H264 stuff hard-coded.
To really announce H265 support to an ONVIF client we would also have some small additions here: https://github.com/nanosonde/sip2rtsp/blob/main/pyonvifsrv/service_media.py (Search for `H264` in this file and you will get an impression what to do)

**Reply by nanosonde** [reply]
Added this issue: https://github.com/nanosonde/sip2rtsp/issues/13

**Comment by nanosonde** [comment]
Another `launch_string` that you could try for H265 without reencoding could be (untested):

`rtspsrc location=rtsp://<IP>:8554/h265_stream name=rtspsrc ! transcodebin profile=video/x-h265 avoid-reencoding=true ! rtph265pay config-interval=1 name=pay0 pt=96`

**Comment by nanosonde** [comment]
Just tried this on the command line:

`gst-launch-1.0 rtspsrc protocols=tcp location=rtsp://<user>:<pwd>@10.10.10.41:554//h265Preview_01_main name=rtspsrc ! parsebin ! rtph265pay config-interval=1 name=pay0 pt=96 ! fakesink`

This could work.

**Comment by nanosonde** [comment]
The corresponding `launch_string` looks like this:

`  launch_string: rtspsrc protocols=tcp location=rtsp://<user>:<pwd>@10.10.10.41:554//h265Preview_01_main name=rtspsrc ! parsebin ! rtph265pay config-interval=1 name=pay0 pt=96`

Without audio so far.

**Comment by nanosonde** [comment]
With audio from SIP: 
Use this in config.yaml:

`launch_string: rtspsrc protocols=tcp location=rtsp://<user>:<pwd>@10.10.10.41:554//h265Preview_01_main name=rtspsrc ! parsebin ! rtph265pay config-interval=1 name=pay0 pt=96 pulsesrc device="BaresipSpeakerInput" do-timestamp=true ! queue ! audioconvert ! audioresample ! audio/x-raw,format=S16LE,channels=1,rate=8000 ! opusenc ! rtpopuspay name=pay1`

**Reply by whitetiger21022014** [reply]
Hello,
Thank for your reply, I like your hard-coded design because it's simple

Used Onvif client tool: ONVIF Device test tool, ONVIF Device Manager
I tested H264 link,  e.g self.streamUri = "rtsp://wowzaec2demo.streamlock.net/vod/mp4:BigBuckBunny_115k.mp4"

I tested your code as onvif server, used only one module (sip2rtsp/pyonvifsrv), it worked with H264 profile so well

With H265 profile, something will add media2's response

https://www.onvif.org/ver10/media/wsdl/media.wsdl
And https://www.onvif.org/ver20/media/wsdl/media.wsdl

http://x.x.x.x/onvif/media_service
And URI: http://x.x.x.x/onvif/media2_service 

Response: the trt tag will change to tr2

**Example**
**GetStream URI**
Response:
HTTP/1.1 200 OK
Connection: close
Content-Length: 529
Content-Type: application/soap+xml; charset=utf-8
CSeq: 0

<?xml version="1.0" encoding="utf-8" standalone="yes"?>
<s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope" xmlns:sc="http://www.w3.org/2003/05/soap-encoding" xmlns:tr2="http://www.onvif.org/ver20/media/wsdl" xmlns:tt="http://www.onvif.org/ver10/schema">
  <s:Body>
    <tr2:GetProfilesResponse>
      <tr2:Profiles token="Profile000" fixed="true">
        <tr2:Name>Profile000</tr2:Name>
      </tr2:Profiles>
      <tr2:Profiles token="Profile001" fixed="true">
        <tr2:Name>Profile001</tr2:Name>
      </tr2:Profiles>
    </tr2:GetProfilesResponse>
  </s:Body>
</s:Envelope>




**GetProfiles**
Response:

HTTP/1.1 200 OK
Connection: close
Content-Length: 529
Content-Type: application/soap+xml; charset=utf-8
CSeq: 0

<?xml version="1.0" encoding="utf-8" standalone="yes"?>
<s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope" xmlns:sc="http://www.w3.org/2003/05/soap-encoding" xmlns:tr2="http://www.onvif.org/ver20/media/wsdl" xmlns:tt="http://www.onvif.org/ver10/schema">
  <s:Body>
    <tr2:GetProfilesResponse>
      <tr2:Profiles token="Profile000" fixed="true">
        <tr2:Name>Profile000</tr2:Name>
      </tr2:Profiles>
      <tr2:Profiles token="Profile001" fixed="true">
        <tr2:Name>Profile001</tr2:Name>
      </tr2:Profiles>
    </tr2:GetProfilesResponse>
  </s:Body>
</s:Envelope>

**Comment by nanosonde** [comment]
Hi @whitetiger21022014 !
Thanks for your reply.

There seems to be one mistake: the example for GetStreamUri shows the GetProfilesResponse and not the GetStreamUriResponse. Could you please show the correct response? That would be helpful.

**Reply by whitetiger21022014** [reply]
Thank you
I attach  IP camera-KBvision's response  (use ONVIF Device test tool), I think onvif server https://www.happytimesoft.com/products/onvif-server/index.html also have the same response

[onvif-media2-response.zip](https://github.com/nanosonde/sip2rtsp/files/11074961/onvif-media2-response.zip)

**Reply by whitetiger21022014** [reply]
**GetStreamUri (/onvif/media2_service )**

**for Profile000 profile (main stream, subtype=0)**
Request:

POST /onvif/media2_service HTTP/1.1
Host: 192.168.1.130
Content-Type: application/soap+xml; charset=utf-8
Authorization: ........

<?xml version="1.0" encoding="utf-8"?>
<s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope">
  <s:Body xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
    <GetStreamUri xmlns="http://www.onvif.org/ver20/media/wsdl">
      <Protocol>RTSP</Protocol>
      <ProfileToken>Profile000</ProfileToken>
    </GetStreamUri>
  </s:Body>
</s:Envelope>

Response:

HTTP/1.1 200 OK
Connection: close
Content-Length: 463
Content-Type: application/soap+xml; charset=utf-8
CSeq: 0

<?xml version="1.0" encoding="utf-8" standalone="yes"?>
<s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope" xmlns:sc="http://www.w3.org/2003/05/soap-encoding" xmlns:tr2="http://www.onvif.org/ver20/media/wsdl" xmlns:tt="http://www.onvif.org/ver10/schema">
  <s:Body>
    <tr2:GetStreamUriResponse>
      <tr2:Uri>rtsp://192.168.1.130:554/cam/realmonitor?channel=1&amp;subtype=0&amp;unicast=true&amp;proto=Onvif</tr2:Uri>
    </tr2:GetStreamUriResponse>
  </s:Body>
</s:Envelope>


**for Profile001 profile (sub stream, subtype=1)**

Request:

POST /onvif/media2_service HTTP/1.1
Host: 192.168.1.130
Content-Type: application/soap+xml; charset=utf-8
Authorization: ....
Content-Length: 368

<?xml version="1.0" encoding="utf-8"?>
<s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope">
  <s:Body xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
    <GetStreamUri xmlns="http://www.onvif.org/ver20/media/wsdl">
      <Protocol>RTSP</Protocol>
      <ProfileToken>Profile001</ProfileToken>
    </GetStreamUri>
  </s:Body>
</s:Envelope>

Response:

HTTP/1.1 200 OK
Connection: close
Content-Length: 463
Content-Type: application/soap+xml; charset=utf-8
CSeq: 0

<?xml version="1.0" encoding="utf-8" standalone="yes"?>
<s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope" xmlns:sc="http://www.w3.org/2003/05/soap-encoding" xmlns:tr2="http://www.onvif.org/ver20/media/wsdl" xmlns:tt="http://www.onvif.org/ver10/schema">
  <s:Body>
    <tr2:GetStreamUriResponse>
      <tr2:Uri>rtsp://192.168.1.130:554/cam/realmonitor?channel=1&amp;subtype=1&amp;unicast=true&amp;proto=Onvif</tr2:Uri>
    </tr2:GetStreamUriResponse>
  </s:Body>
</s:Envelope>

**Comment by whitetiger21022014** [comment]
And GetStreamUri
[media2 - GetStreamUri - correct.txt](https://github.com/nanosonde/sip2rtsp/files/11075069/media2.-.GetStreamUri.-.correct.txt)
